MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -141,6 +151,98 @@ RemoveMemoryMapEntry (
   }
 }
 
+/**
+  Helper function to evaluate if memory regions intersect.
+
+  @param  Start1     The address of the first byte in the first memory region.
+  @param  End1       The address of the last byte in the first memory region.
+  @param  Start2     The address of the first byte in the second memory region.
+  @param  End2       The address of the last byte in the second memory region.
+
+  @return TRUE if the memory regions intersect, FALSE otherwise.
+**/
+STATIC
+BOOLEAN
+MemoryRegionsIntersect (
+  IN EFI_PHYSICAL_ADDRESS  Start1,
+  IN EFI_PHYSICAL_ADDRESS  End1,
+  IN EFI_PHYSICAL_ADDRESS  Start2,
+  IN EFI_PHYSICAL_ADDRESS  End2
+  )
+{
+  return (((Start1 <= End2) && (Start2 <= Start1)) ||
+          ((Start2 <= End1) && (Start1 <= Start2)));

```

Comment:
This can be simplified to `Start1 < End2 && Start2 < End1` but we should add an ASSERT() that Start1 <= End1 and Start2 <= End2

Reply:
Updated.